Adds abstract test classes for serialisation

colings86 opened the pull request:
This adds test classes that can be used to test the wire serialisation and (optionally) the XContent serialisation of objects that implement Streamable/Writeable and ToXContent.

These test classes will enable classes such as InternalAggregation (or at least its implementations) to be tested in a consistent way when is comes to testing serialisation.

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableTestCase.java:
```diff
@@ -0,0 +1,98 @@
+/*
+ * ELASTICSEARCH CONFIDENTIAL
```
wrong header

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableTestCase.java:
```diff
@@ -0,0 +1,98 @@
+import static org.hamcrest.Matchers.equalTo;
+
+public abstract class AbstractStreamableTestCase<T extends Streamable> extends ESTestCase {
+    protected static final int NUMBER_OF_TESTQUERIES = 20;
```
why don't you make it random ie instead of this constant use `randomIntBetween(1, 20)`

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableTestCase.java:
```diff
@@ -0,0 +1,98 @@
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            instance.writeTo(output);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(),
+                    new NamedWriteableRegistry(Collections.emptyList()))) {
```
will this work for all serialization classes? I mean queries need to be registered etc. I think we have to add all the ones we need for the subclass. So maybe add a factory method we can override?

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java:
```diff
@@ -0,0 +1,95 @@
+/*
+ * ELASTICSEARCH CONFIDENTIAL
```
you should run `gradle precommit`

s1monw commented on test/framework/src/main/java/org/elasticsearch/test/AbstractWireSerializingTestCase.java:
```diff
@@ -0,0 +1,95 @@
+        try (BytesStreamOutput output = new BytesStreamOutput()) {
+            instance.writeTo(output);
+            try (StreamInput in = new NamedWriteableAwareStreamInput(output.bytes().streamInput(),
+                    new NamedWriteableRegistry(Collections.emptyList()))) {
```
same comment as above

jpountz commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableTestCase.java:
```diff
@@ -0,0 +1,102 @@
+                    equalTo(firstInstance.hashCode()));
+
+            T secondInstance = copyInstance(firstInstance);
+            assertTrue("query is not equal to self", secondInstance.equals(secondInstance));
```
maybe use assertEquals instead to have better error messages

jpountz commented on test/framework/src/main/java/org/elasticsearch/test/AbstractStreamableXContentTestCase.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Subclasses can override this method and return an array of fieldnames
+     * which should be protected from recursive random shuffling in the
+     * {@link #testFromXContent()} test case
```
can the docs give an example when it is useful to do so?

colings86 commented:
@s1monw @jpountz I pushed commits addressing your comments. I have implemented InternalMinTests which uses `AbstractWireSerializingTestCase`. I plan in a follow up PR to convert `AbstractQueryTestCase` and `BaseAggregationTestCase` to extend `AbstractSerializingTestCase`. Not sure if it's worth keeping the `Streamable` versions of the test cases since we want to move everything to use `Writeable` anyway?

colings86 commented:
Also worth noting that because the test cases rely on `hashCode` and `equals` being implemented on the test class I had to add implementations to `InternalAggregation` and `InternalNumericMetricAggregation` which both declare methods that their subcleasses need to implement. In order to keep this PR mamageable and not require that I implemented `hashCode` and `equals` for all `InternalAggregation`s subclasses I added a default implementation of those declared methods which use object identity, as well as adding a `norelease` comment to them stating that they should be made abstract when all the `InternalAggregation` subclasses implement the methods

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/InternalAggregation.java:
```diff
@@ -189,6 +190,36 @@ public final XContentBuilder toXContent(XContentBuilder builder, Params params)
+    }
+
+    // norelease: make this abstract when all InternalAggregations implement this method
+    protected boolean doEquals(Object obj) {
```
Can you add javadocs?

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalNumericMetricsAggregation.java:
```diff
@@ -102,4 +103,25 @@ private InternalNumericMetricsAggregation(String name, List<PipelineAggregator>
+    // norelease: make this abstract when all InternalAggregations implement this method
+    protected boolean innerEquals(InternalNumericMetricsAggregation obj) {
+        return this == obj;
+    }
```
I don't like having both doEquals and innerEquals. Maybe just override equals to also include the format and still expect sub classes to implement doEquals? (Same for the hashcode)

colings86 commented:
@jpountz I pushed a new commit addressing your latest comments. Could you review again?
